test/dist_autograd_test.py

```diff
@@ -1292,5 +1292,77 @@ def test_backwards_nested_python_udf(self):
             self.assertEqual(t1.grad, grads[t1])
             self.assertEqual(t2.grad, grads[t2])
 
+
+    _test_clean_context_backward_context_id = None
+
+    class MyBackwardFunc(Function):
+        @staticmethod
+        def forward(ctx, input):
+            return input
+
+        @staticmethod
+        @once_differentiable
+        def backward(ctx, input):
+            assert(DistAutogradTest._test_clean_context_backward_context_id is not None)
+
+            # Release the context to simulate error (use barrier before releasing context to ensure all nodes execute the backward function).
+            dist.barrier()
+            dist_autograd._release_context(DistAutogradTest._test_clean_context_backward_context_id)
+
+            # Verify all contexts are cleaned up.
+            assert(_all_contexts_cleaned_up())
+
+            return input
+
+    @dist_init
+    def test_clean_context_during_backward(self):
+        '''
+        This test simulates the situation where the 'backward' call might throw
+        an exception locally which would lead to the autograd context being
+        cleaned up if we're using the context manager. As a result, the autograd
+        context might be cleaned up while some threads are still using the
+        autograd context.
+
+        It is fine for the 'backward' call to throw an exception in this test,
+        but the process should not crash.
+        '''
+        self._initialize_pg()
+
+        context = dist_autograd._new_context()
+        context_id = context._context_id()
+        DistAutogradTest._test_clean_context_backward_context_id = context_id
+
+        # Send the context id to all nodes.
+        for i in range(0, self.world_size):
+            if i != self.rank:
+                rpc.rpc_sync("worker{}".format(i), _set_rpc_done, args=(context_id, 1))
+
+        dist.barrier()
+
+        # Verify all context ids have been received.
+        self.assertEqual(self.world_size - 1, len(known_context_ids))
+
+        t1 = torch.rand((3, 3), requires_grad=True)
+        for i in range(0, 100):
+            dst = self._next_rank()
+            t1 = rpc.rpc_sync("worker{}".format(dst), torch.add, args=(t1, t1))
+            if i == 99:
+                # Call MyBackwardFunc as the first op of the backward pass to
+                # ensure we release the context early in the backward pass.
+                t1 = DistAutogradTest.MyBackwardFunc.apply(t1)
+        self.assertEqual(100, len(context._send_functions()))
+
+        with self.assertRaisesRegex(RuntimeError, "Could not find autograd context with id"):
+            dist_autograd.backward([t1.sum()])
+
+        # HACK: Killing workers since otherwise the autograd engine gets stuck on
+        # other nodes. The proper fix would be addressing:
```

Comment:
Let's be more specific on why it might stuck on others. 

Would I be correct if I assume it stuck because the crashing backward destroyed the context on this node, and hence the `next_rank` won't be able to clear the context when exiting the scope.